internal/providers/azure/azure.go

```diff
@@ -56,15 +57,38 @@ const (
 	CDS_FSTYPE_UDF = "udf"
 )
 
-// FetchConfig wraps FetchOvfDevice to implement the platform.NewFetcher interface.
+var (
+	imdsURL = url.URL{
+		Scheme: "http",
+		Host:   "169.254.169.254",
+		Path:   "metadata/instance/compute/userData?api-version=2021-01-01&format=text",
+	}
+)
+
+// FetchConfig wraps FetchFromAzureMetadata to implement the platform.NewFetcher interface.
 func FetchConfig(f *resource.Fetcher) (types.Config, report.Report, error) {
-	return FetchFromOvfDevice(f, []string{CDS_FSTYPE_UDF})
+	return FetchFromAzureMetadata(f, []string{CDS_FSTYPE_UDF})
 }
 
-// FetchFromOvfDevice has the return signature of platform.NewFetcher. It is
+// FetchFromAzureMetadata has the return signature of platform.NewFetcher. It is
 // wrapped by this and AzureStack packages.
-func FetchFromOvfDevice(f *resource.Fetcher, ovfFsTypes []string) (types.Config, report.Report, error) {
+// It first tries to fetch user-data from Azure IMDS and in case of failures, it fallbacks to custom-data.
+func FetchFromAzureMetadata(f *resource.Fetcher, ovfFsTypes []string) (types.Config, report.Report, error) {
+	// fetch-offline is not supported since we first try to get config from Azure IMDS.
+	// this config fetching can only run in fetch stage.
+	if f.Offline {
+		return types.Config{}, report.Report{}, resource.ErrNeedNet
+	}
+
 	logger := f.Logger
+
+	// we first try to config from IMDS, in case of failure we fallback on the custom data.
+	data, err := f.FetchToBuffer(imdsURL, resource.FetchOptions{Compression: "base64"})
+	if err == nil {
```

Comment:
If there's any error, we're falling through and failing to report it.  I assume the metadata service is always available, and returns 404 if userdata is missing?  If so, we should probably fall through on `resource.ErrNotFound`, and error out otherwise.

Reply:
Just checked and if userdata is missing it seems to return 200 with an empty string:
```
curl -H Metadata:true --noproxy "*" "http://169.254.169.254/metadata/instance/compute/userData?api-version=2021-01-01&format=text" -v
*   Trying 169.254.169.254:80...
* Connected to 169.254.169.254 (169.254.169.254) port 80 (#0)
> GET /metadata/instance/compute/userData?api-version=2021-01-01&format=text HTTP/1.1
> Host: 169.254.169.254
> User-Agent: curl/7.79.1
> Accept: */*
> Metadata:true
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Content-Type: text/plain; charset=utf-8
< Server: IMDS/150.870.65.597
< Date: Thu, 14 Apr 2022 09:07:49 GMT
< Content-Length: 0
< 
* Connection #0 to host 169.254.169.254 left intact
``` 

For any error in the `FetchToBuffer`, we log them the line after:
```go
logger.Debug("failed to retrieve user-data from IMDS, falling back to custom-data: %v", err)
```

isnt' enough ?